tips/TIP-0018/tip-0018.md

```diff
@@ -0,0 +1,3973 @@
+  - <i>Issuer Block</i>
+  - <i>Metadata Block</i>
+- <i>Blocks</i> must be sorted in ascending order based on their `Block Type` both in `Blocks` and `Immutable Blocks`
+  fields.
+- Syntactic validation of all present feature blocks must pass.
+- When `Alias ID` is zeroed out, `State Index` and `Foundry Counter` must be `0`.
+- `State Metadata Length` must not be greater than `Max Metadata Length`.
+- `Address` of <i>State Controller Address Unlock Condition</i> and `Address` of
+  <i>Governor Address Unlock Condition</i> must be different from the alias address derived from `Alias ID`.
+
+### Additional Transaction Semantic Validation Rules
+
+- Explicit `Alias ID`: `Alias ID` is taken as the value of the `Alias ID` field in the alias output.
+- Implicit `Alias ID`: When an alias output is consumed as an input in a transaction and `Alias ID` field is zeroed out
+  while `State Index` and `Foundry Counter` are zero, take the BLAKE2b-160 hash of the `Output ID` of the input as
+  `Alias ID`.
+- The BLAKE2b-160 hash function outputs a 20 byte hash as opposed to the 32 byte hash size of BLAKE2b-256. `Alias ID`
+  is sufficiently secure and collision free with 20 bytes already, as it is actually a hash of an `Output ID`, which
+  already contains the 32 byte hash `Transaction ID`.
+- For every non-zero explicit `Alias ID` on the output side there must be a corresponding alias on the input side. The
+  corresponding alias has the explicit or implicit `Alias ID` equal to that of the alias on the output side.
+
+#### Consumed Outputs
+
+Whenever an alias output is consumed in a transaction, it means that the alias is transitioned into its next state. The
+**current state** is defined as the **consumed alias output**, while the **next state** is defined as the **alias
+output with the same explicit `AliasID` on the output side**. There are two types of transitions: `state transition`
+and `governance transition`.
+- State transition:
+    - A state transition is identified by an incremented `State Index`.
+    - The `State Index` must be incremented by 1.
+    - The unlock block must correspond to the `Address` of <i>State Controller Address Unlock Condition</i>.
+    - State transition can only change the following fields in the next state:
+      - `IOTA Amount`,
+      - `Native Tokens`,
+      - `State Index`,
+      - `State Metadata Length`,
+      - `State Metadata`,
+      - `Foundry Counter` and
+      - `Sender Block` in `Blocks`.
+    - `Foundry Counter` field must increase by the number of foundry outputs created in the transaction that map to
+      `Alias ID`. The `Serial Number` fields of the created foundries must be the set of natural numbers that cover the
+       open-ended interval between the previous and next values of the `Foundry Counter` field in the alias output.
+    - The created foundry outputs must be sorted in the list of outputs by their `Serial Number`. Note, that any
```

Comment:
More generally, does `sorted` here means that nothing can be between them ?

Reply:
No, `sorted` here means that they have to be sorted wrt each other. Then we can check monotonicity and uniqueness easy while iterating over the list of outputs from the beginning.

Reply:
So just to be sure, something like this would be fine? : Created Foundry (6), Basic, Transitioned Foundry (4), Alias, Created Foundry (7), Created Foundry (8), Basic

Reply:
Yes, this would be fine, assuming InputAlias(FoundryCounter) = 5 and outputAlias(FoundryCounter) = 8

Reply:
Thanks 👍